client/consensus/common/src/lib.rs

```diff
@@ -74,47 +83,100 @@ impl<B: BlockT> ParachainConsensus<B> for Box<dyn ParachainConsensus<B> + Send +
+		};
+
+		let monitor =
+			level_limit.map(|level_limit| SharedData::new(LevelMonitor::new(level_limit, backend)));
+
+		Self { inner, monitor }
 	}
 }
 
-impl<I: Clone> Clone for ParachainBlockImport<I> {
+impl<Block: BlockT, I: Clone, BE> Clone for ParachainBlockImport<Block, I, BE> {
 	fn clone(&self) -> Self {
-		ParachainBlockImport(self.0.clone())
+		ParachainBlockImport { inner: self.inner.clone(), monitor: self.monitor.clone() }
 	}
 }
 
 #[async_trait::async_trait]
-impl<Block, I> BlockImport<Block> for ParachainBlockImport<I>
+impl<Block, BI, BE> BlockImport<Block> for ParachainBlockImport<Block, BI, BE>
 where
 	Block: BlockT,
-	I: BlockImport<Block> + Send,
+	BI: BlockImport<Block> + Send,
+	BE: Backend<Block>,
 {
-	type Error = I::Error;
-	type Transaction = I::Transaction;
+	type Error = BI::Error;
+	type Transaction = BI::Transaction;
 
 	async fn check_block(
 		&mut self,
 		block: sc_consensus::BlockCheckParams<Block>,
 	) -> Result<sc_consensus::ImportResult, Self::Error> {
-		self.0.check_block(block).await
+		self.inner.check_block(block).await
 	}
 
 	async fn import_block(
 		&mut self,
-		mut block_import_params: sc_consensus::BlockImportParams<Block, Self::Transaction>,
+		mut params: sc_consensus::BlockImportParams<Block, Self::Transaction>,
 		cache: std::collections::HashMap<sp_consensus::CacheKeyId, Vec<u8>>,
 	) -> Result<sc_consensus::ImportResult, Self::Error> {
+		// Blocks are stored within the backend by using POST hash.
+		let hash = params.post_hash();
+		let number = *params.header.number();
+
 		// Best block is determined by the relay chain, or if we are doing the initial sync
 		// we import all blocks as new best.
-		block_import_params.fork_choice = Some(sc_consensus::ForkChoiceStrategy::Custom(
-			block_import_params.origin == sp_consensus::BlockOrigin::NetworkInitialSync,
+		params.fork_choice = Some(sc_consensus::ForkChoiceStrategy::Custom(
+			params.origin == sp_consensus::BlockOrigin::NetworkInitialSync,
 		));
-		self.0.import_block(block_import_params, cache).await
+
+		let maybe_lock = self.monitor.as_ref().map(|monitor_lock| {
+			let mut monitor = monitor_lock.shared_data_locked();
+			monitor.enforce_limit(number);
+			monitor.release_mutex()
+		});
+
+		let res = self.inner.import_block(params, cache).await?;
+
+		if let (Some(mut monitor_lock), ImportResult::Imported(_)) = (maybe_lock, &res) {
+			let mut monitor = monitor_lock.upgrade();
+			monitor.block_imported(number, hash);
+		}
+
+		match res {
+			ImportResult::Imported(_) =>
+				log::debug!(target: "parachain", "XXX Imported (@{}) {}", number, hash),
+			_ =>
+				log::warn!(target: "parachain", "XXX Failed importing (@{}) {}: {:?}", number, hash, res),
+		}
+
+		Ok(res)
 	}
 }
+
+/// Marker trait denoting a block import type that fits the parachain requirements.
+pub trait ParachainBlockImportMarker {}
+
+impl<B: BlockT, BI, BE> ParachainBlockImportMarker for ParachainBlockImport<B, BI, BE> {}
```

Comment:
This should probably be sealed.

Reply:
Ok. I'll make it sealed.

What do you think about renaming:
```
    struct ParachainBlockImport to ParachainBasicBlockImport
    trait ParachainBlockImportMarker to ParachainBlockImport
```

And then use a "supertrait" for `ParachainBlockImport` trait:

```
trait ParachainBlockImport<...>: BlockImport<...> {}
```

---

As I said the advantage of doing so would be that the users can do:

```
BI: ParachainBlockImport<B>
```

While currently (with the simple marker) they should do:

```
BI: BlockImport<B> + ParachainBlockImportMarker
```

Reply:
Supertrait sounds like a good idea. The naming. IDK. Naming is hard :P "Basic" sounds weird TBH.

Reply:
This was not done yet or @davxy?

Reply:
No, not yet. I would be happy to do this before merge but I've been held back by the *naming dilemma*.

To me as the general trait to import blocks is `BlockImport` makes sense to rename the parachain one as `ParachainBlockImport`.

But then we have to find a proper name for the provided implementation :-D
Indeed `BasicParachainBlockImport` sounds weird to me as well but I can't find something better...

We have to find a proper prefix or leave it as is (and use a marker, that in the end is not so bad)

Reply:
```
struct ParachainBlockImport;

trait ParachainBlockImportMarker
```

Still sounds good to me.